server/middleware/course-instances.coffee

```diff
@@ -21,93 +23,104 @@ module.exports =
       { _id: { $in: userObjectIDs } },
       { $addToSet: { courseInstances: courseInstance._id } }
     )
-    
+
     res.status(200).send(courseInstance.toObject({ req }))
 
 
   fetchNextLevel: wrap (req, res) ->
     levelOriginal = req.params.levelOriginal
-    if not database.isID(levelOriginal)
-      throw new errors.UnprocessableEntity('Invalid level original ObjectId')
-    
+    unless database.isID(levelOriginal) then throw new errors.UnprocessableEntity('Invalid level original ObjectId')
+    sessionID = req.params.sessionID
+    unless database.isID(sessionID) then throw new errors.UnprocessableEntity('Invalid session ObjectId')
     courseInstance = yield database.getDocFromHandle(req, CourseInstance)
-    if not courseInstance
-      throw new errors.NotFound('Course Instance not found.')
+    unless courseInstance then throw new errors.NotFound('Course Instance not found.')
+    classroom = yield Classroom.findById courseInstance.get('classroomID')
+    unless classroom then throw new errors.NotFound('Classroom not found.')
+    currentLevel = yield Level.findOne({original: mongoose.Types.ObjectId(levelOriginal)}, {practiceThresholdMinutes: 1, type: 1})
+    unless currentLevel then throw new errors.NotFound('Current level not found.')
+
     courseID = courseInstance.get('courseID')
+    courseLevels = []
+    courseLevels = course.levels for course in classroom.get('courses') or [] when courseID.equals(course._id)
 
-    classroom = yield Classroom.findById courseInstance.get('classroomID')
-    if not classroom
-      throw new errors.NotFound('Classroom not found.')
-      
-    nextLevelOriginal = null
-    foundLevelOriginal = false
-    for course in classroom.get('courses') or []
-      if not courseID.equals(course._id)
-        continue
-      for level, index in course.levels
-        if level.original.toString() is levelOriginal
-          foundLevelOriginal = true
-          nextLevelOriginal = course.levels[index+1]?.original
-          break
-          
-    if not foundLevelOriginal
-      throw new errors.NotFound('Level original ObjectId not found in Classroom courses')
-    
-    if not nextLevelOriginal
-      return res.status(200).send({})
-      
+    # Get level completions and playtime
+    currentLevelSession = null
+    levelIDs = (level.original.toString() for level in courseLevels)
+    query = {$and: [{creator: req.user.id}, {'level.original': {$in: levelIDs}}]}
+    levelSessions = yield LevelSession.find(query, {level: 1, playtime: 1, state: 1})
+    levelCompleteMap = {}
+    for levelSession in levelSessions
+      currentLevelSession = levelSession if levelSession.id is sessionID
+      levelCompleteMap[levelSession.get('level')?.original] = levelSession.get('state')?.complete
+    unless currentLevelSession then throw new errors.NotFound('Level session not found.') 
+    needsPractice = Math.round(currentLevelSession.get('playtime') / 60) > (currentLevel.get('practiceThresholdMinutes') ? 0)
```

Comment:
This is wrong, I'll fix it later today. No threshold should not default to zero.